src/python.cc

```diff
@@ -1468,13 +1488,76 @@ ModelInstanceState::ResponseSendDecoupled(
         InferResponse::LoadFromSharedMemory(
             shm_pool_, send_message_payload->response,
             false /* open cuda ipc handle */);
-    TRITONSERVER_Error* error = infer_response->Send(
-        response_factory, CudaStream(), send_message_payload->flags);
+
+    bool requires_deferred_callback = false;
+    std::vector<std::pair<std::unique_ptr<PbMemory>, void*>> gpu_output_buffers;
+    std::shared_ptr<TRITONSERVER_Error*> error = infer_response->Send(
+        response_factory, CudaStream(), requires_deferred_callback,
+        send_message_payload->flags, shm_pool_, gpu_output_buffers);
     SetErrorForResponseSendMessage(send_message_payload, error, error_message);
+
+    if (requires_deferred_callback) {
+      AllocatedSharedMemory<char> gpu_buffers_handle =
+          shm_pool_->Construct<char>(
+              sizeof(uint64_t) +
+              gpu_output_buffers.size() *
+                  sizeof(bi::managed_external_buffer::handle_t));
+      uint64_t* gpu_buffer_count =
+          reinterpret_cast<uint64_t*>(gpu_buffers_handle.data_.get());
+      *gpu_buffer_count = gpu_output_buffers.size();
+      bi::managed_external_buffer::handle_t* gpu_buffers_handle_shm =
+          reinterpret_cast<bi::managed_external_buffer::handle_t*>(
+              gpu_buffers_handle.data_.get() + sizeof(uint64_t));
+      send_message_payload->gpu_buffers_handle = gpu_buffers_handle.handle_;
+
+      size_t index = 0;
+      for (auto& output_buffer_pair : gpu_output_buffers) {
+        std::unique_ptr<PbMemory>& pb_memory = output_buffer_pair.first;
+        gpu_buffers_handle_shm[index] = pb_memory->ShmHandle();
+        ++index;
+      }
+
+      // Additional round trip so that the stub can fill the GPU output buffers.
```

Comment:
Why there is another round trip? the backend process signals back that there is CUDA IPC handle and wait for stub process to copy the data to within the CUDA IPC handle?

Do I summarize the workflow correctly:

1. stub requests for output buffer (and passing buffer via shared memory at the same time if CPU tensor)
2. backend acquires output buffer, copy if CPU tensor, otherwise, passing the CUDA IPC handle back to stub.
3. (only for GPU tensor) stub then copy GPU tensor into the CUDA IPC handle

Reply:
Exactly. The third round is what requires the round trip.